### PR TITLE
fix #486 - prevent users to share table with themselves

### LINKED ISSUE #486: Prevent users to share their tables with themselves
### Current behavior

Users can share their tables with themselves. They can share by writing their user names to share box or they can share by sharing with their organizations.
 
### Expected behavior
Users can't share their tables with themselves. If they share table with organization level then they still can't see their tables on the 'Data Tables Shared with you' section.

### Steps to reproduce the behavior

1. GIVEN user has a table they want to share
2. WHEN they search individual members at the shared box on the Silo edit page
3. THEN they can see and select their names

1. GIVEN user has a table they want to share
2. WHEN they check the 'share with organization button' and click the save
3. THEN they see that table on their 'Data Tables Shared with you' section on the silo list page.

### Additional details
Environment: [**dev**/**staging**/**demo**/**production**]

### PR DESCRIPTION
**Pull request by yilmazbekdemir**
## Purpose
if users share their own tables with themselves or their organizations, they can see same table at the 'Tables shared with you section'

## Approach
Edited query for shared_silos on silo list page.

_Related ticket: #486

### CODE REVIEW

File: silo/tests/test_views.py [hunk 1]
```diff
@@ -1307,4 +1307,45 @@ def test_list_public_silo(self):
...
+                              owner=self.tola_user.user)
+
+        silo = factories.Silo(name='Test Share Silo',
+                              owner=self.tola_user.user,
+                              reads=[read],
+                              public=False,
+                              shared=[self.tola_user.user],
+                              share_with_organization=False)
+
+        url = reverse('list_silos')
```
Comment (jefmoura) [inline silo/tests/test_views.py]: You don't need to use the reverse function. Just use the view.

File: silo/tests/test_views.py [hunk 2]
```diff
@@ -1307,4 +1307,45 @@ def test_list_public_silo(self):
...
+
+        request = self.factory.get(url)
+        request.user = self.tola_user.user
+        response = views.list_silos(request)
+        match = '<a href="/silo_edit/%s">Test Share Silo</a>' % silo.pk
+        print(match)
```
Comment (jefmoura) [inline silo/tests/test_views.py]: Delete this print

File: silo/views.py [hunk 3]
```diff
@@ -933,12 +933,13 @@ def list_silos(request):
     shared_silos = Silo.objects.filter(Q(shared__id=user.pk) |
                                        Q(share_with_organization=True,
                                          owner__tola_user__organization=\
-                                         user.tola_user.organization)
-                                       ).prefetch_related("reads")
+                                         user.tola_user.organization))\
```
Comment (jefmoura) [inline silo/views.py]: This is not in PEP8

File: silo/tests/test_views.py [hunk 4]
```diff
@@ -1178,45 +1178,41 @@ def setUp(self):
 
     def test_list_silos_share_with_users_organization(self):
         request_user = factories.User(username='Another User')
-        request_tola_user = factories.TolaUser(user=request_user,
+        factories.TolaUser(user=request_user,
```
Comment (jefmoura) [inline silo/tests/test_views.py]: This isn't in PEP8

File: silo/tests/test_views.py [hunk 5]
```diff
@@ -1178,45 +1178,41 @@ def setUp(self):
...
                                                organization=self.tola_user
                                                                 .organization)
 
         read = factories.Read(read_name="test_data",
                               owner=self.tola_user.user)
 
-        silo = factories.Silo(name='Test Share Silo',
+        factories.Silo(name='Test Share Silo',
```
Comment (jefmoura) [inline silo/tests/test_views.py]: This isn't in PEP8

File: silo/tests/test_views.py [hunk 6]
```diff
@@ -1178,45 +1178,41 @@ def setUp(self):
...
+        request = self.factory.get('')
         request.user = request_user
         response = views.list_silos(request)
         self.assertEqual(response.status_code, 200)
         self.assertContains(response, 'Test Share Silo')
 
     def test_list_silos_share_with_different_organization(self):
         request_user = factories.User(username='Another User')
-        request_tola_user = factories.TolaUser(user=request_user)
+        factories.TolaUser(user=request_user)
```
Comment (jefmoura) [inline silo/tests/test_views.py]: This isn't in PEP8

File: silo/tests/test_views.py [hunk 7]
```diff
@@ -1178,45 +1178,41 @@ def setUp(self):
...
 
         read = factories.Read(read_name="test_data",
                               owner=self.tola_user.user)
 
-        silo = factories.Silo(name='Test Share Silo',
+        factories.Silo(name='Test Share Silo',
```
Comment (jefmoura) [inline silo/tests/test_views.py]: This isn't in PEP8

File: silo/tests/test_views.py [hunk 8]
```diff
@@ -1231,16 +1227,14 @@ def test_list_silos_not_share_with_organization(self):
...
```
Comment (jefmoura) [inline silo/tests/test_views.py]: This isn't in PEP8 and the var `request_tola_user` isn't used.

File: silo/tests/test_views.py [hunk 9]
```diff
@@ -1251,16 +1245,14 @@ def test_list_silos_with_owner_user(self):
...
```
Comment (jefmoura) [inline silo/tests/test_views.py]: This isn't in PEP8

File: silo/tests/test_views.py [hunk 10]
```diff
@@ -1272,16 +1264,14 @@ def test_list_silos_with_shared_user(self):
...
```
Comment (jefmoura) [inline silo/tests/test_views.py]: This isn't in PEP8

File: silo/tests/test_views.py [hunk 11]
```diff
@@ -1294,17 +1284,51 @@ def test_list_public_silo(self):
...
```
Comment (jefmoura) [inline silo/tests/test_views.py]: This isn't in PEP8

File: silo/views.py [hunk 12]
```diff
@@ -927,22 +927,24 @@ def list_silos(request):
...
-    shared_silos = Silo.objects.filter(Q(shared__id=user.pk) |
-                                       Q(share_with_organization=True,
-                                         owner__tola_user__organization=\
-                                         user.tola_user.organization)
-                                       ).prefetch_related("reads")
+    shared_silos = Silo.objects.filter(
+                Q(shared__id=user.pk) |
+                Q(share_with_organization=True,
+                  owner__tola_user__organization=user.tola_user.organization
+                  )
```
Comment (jefmoura) [inline silo/views.py]: Move this bracket to the line above.

File: silo/views.py [hunk 13]
```diff
@@ -927,22 +927,24 @@ def list_silos(request):
...
+            ).exclude(owner=user).prefetch_related("reads")
 
     public_silos = Silo.objects.filter(Q(public=True)
-                                       & ~Q(owner=user)).\
-                                prefetch_related("reads")
+                                       & ~Q(owner=user))\
```
Comment (jefmoura) [inline silo/views.py]: This isn't easy to read as well.

**Comment by jefmoura** [comment]
@yilmazbekdemir You solved only one of the problems.
- [x] Don't show the table twice in the silo list page when you're the owner and shared with the org
- [ ] Don't allow the owner to share the table with himself

**Comment by yilmazbekdemir** [comment]
I thought about this. Owner can share table with his organizations. So, anyway user can share with themselves. Removing username from user search box doesn't solve this but i can check that part also.

**Comment by jefmoura** [comment]
- Sharing with the organization means that the user is sharing with the whole company, not with a specific user.
- Of course, the solution to the problem isn't only removing the user from the search box, you need to validate it.

**Pull request closed without merging**